### PR TITLE
Additions for evolutions 3

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Adds a type alias for the Jacobian type, similar to the `InverseJacobian` type alias
- Fixes a bug in `compute_item_result_impl` that caused really confusing template errors
- Adds tags for the logical coordinates (item) and inverse jacobian (compute item)
- Generalizes the spatial derivative prefix `d` to be a compute item and item depending on its template parameters
- Renames `Tags::d` to `Tags::deriv`
- Fixes a clang-tidy warning in DataBox.hpp

These are again changes I had to make for getting evolutions working.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

File: src/DataStructures/Tensor/TypeAliases.hpp [hunk 1]
```diff
@@ -334,11 +334,15 @@ using ijaa = Tensor<DataType, tmpl::integral_list<std::int32_t, 3, 2, 1, 1>,
...
-
 }  // namespace tnsr
 
 template <size_t Dim, typename Frame1, typename Frame2>
 using InverseJacobian =
     Tensor<DataVector, tmpl::integral_list<std::int32_t, 2, 1>,
            typelist<SpatialIndex<Dim, UpLo::Up, Frame1>,
                     SpatialIndex<Dim, UpLo::Lo, Frame2>>>;
+
+template <size_t Dim, typename Frame1, typename Frame2>
```
Comment (kidder) [inline src/DataStructures/Tensor/TypeAliases.hpp]: I think we should use SourceFrame and TargetFrame as the template parameters (and above in InverseJacobian) to match CoordinateMap

File: src/DataStructures/DataBoxTag.hpp [hunk 2]
```diff
@@ -414,7 +423,7 @@ struct is_variables_compute_item<
 template <template <typename...> class Wrapper, typename TagList,
           typename... Args>
 using wrap_tags_in =
-    tmpl::transform<TagList, tmpl::bind<Wrapper, tmpl::_1, Args...>>;
+    tmpl::transform<TagList, tmpl::bind<Wrapper, tmpl::_1, tmpl::pin<Args>...>>;
```
Comment (kidder) [inline src/DataStructures/DataBoxTag.hpp:432]: is this related to the commit message?

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 3]
```diff
@@ -14,32 +16,122 @@
...
+ * \ingroup DataBoxTagsGroup
+ * \brief Prefix for a spatial derivative
+ *
+ * There are four variants of this tag that change how the derivatives are
+ * computed depending on which is chosen. The two simplest are the non-compute
+ * item versions for Tensor and Variables tags. These take three template
+ * parameters:
+ * 1. The tag to wrap
+ * 2. The volume dim as a type (e.g. `tmpl::size_t<Dim>`)
+ * 3. The frame of the derivative index is in `Tensor`s
```
Comment (kidder) [inline src/DataStructures/DataBox/Prefixes.hpp]: index is in -> index in

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 4]
```diff
@@ -14,32 +16,122 @@
...
+ *
+ * The partial derivatives in a Frame that is not the logical frame as a compute
+ * item is requested by specifying the list of tags in the
+ * `Tags::Variables<tmpl::list<Tags...>>`, the list of tags to be
+ * differentiated, and the Tag for inverse Jacobian between the logical frame
+ * and the frame in which the derivative is taken. Note that the derivative Tags
+ * must be a contiguous subset from the first tag in the `Tags...` pack.
```
Comment (kidder) [inline src/DataStructures/DataBox/Prefixes.hpp]: from the head of the `Tags...` pack

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 5]
```diff
@@ -14,32 +16,122 @@
...
-template <typename Tag, typename VolumeDim, typename Fr>
-struct d<Tag, VolumeDim, Fr, Requires<tt::is_a_v<Tensor, db::item_type<Tag>>>>
-    : db::DataBoxPrefix {
+class DataVector;
+template <size_t Dim>
+class Index;
+template <class>
+class Variables;
+
+template <typename DerivativeTags, typename VariableTags, size_t Dim>
```
Comment (kidder) [inline src/DataStructures/DataBox/Prefixes.hpp]: Having to include these forward declarations suggests to me that we might want to move the compute item tags to where the function is.  Thoughts?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/DataStructures/Tensor/TypeAliases.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/DataStructures/Tensor/TypeAliases.hpp]: I agree, `Frame1` -> `SourceFrame` and `Frame2` -> `TargetFrame`?

File: src/DataStructures/DataBoxTag.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/DataStructures/DataBoxTag.hpp:432]: Indirectly, yes. I'll split it into a separate commit

File: src/DataStructures/DataBox/Prefixes.hpp (hunk 5 above)
Comment (nilsdeppe) [reply src/DataStructures/DataBox/Prefixes.hpp]: I think there are sort of two options for other locations:

1. A tag file closer to the function, but we still forward declare the functions. This will reduce annoying compilation dependencies that will otherwise crop up. That is, if you only need the Tag but don't need the function definition you shouldn't need the function's header (for function templates only is this a concern)

2. We add the Tag to the header of the function. For functions where we have explicit instantiations and the header file does not contain a definition this is probably ideal. For function templates where we can't generate explicit instantiations this could be less optimal.

Have said all that, only functions that act on Variables will we not be able to generate explicit instantiations trivially. However, there aren't too many of these, and what we can do is have the definition in a cpp file, that cpp file then gets included in a cpp file for the system, which generates the explicit instantiations and all problems are solved. 2 + generating instantiations for specific systems by putting the responsibility on that system's author is best.

**Comment by nilsdeppe** [comment]
I've addressed Larry's requests. The outstanding one from his is what to do with compute item tags, which I left a long reply to.

File: src/DataStructures/DataBox/Prefixes.hpp (hunk 5 above)
Comment (kidder) [reply src/DataStructures/DataBox/Prefixes.hpp]: I think I like the last suggestion best; but I'm happy to hear other opinions

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 6]
```diff
@@ -14,32 +16,123 @@
...
+template <size_t Dim>
+class Extents;
+}  // namespace Tags
 
-template <typename Tag, typename VolumeDim, typename Fr>
-struct d<Tag, VolumeDim, Fr, Requires<tt::is_a_v<Tensor, db::item_type<Tag>>>>
+class DataVector;
+template <size_t Dim>
+class Index;
+template <class>
```
Comment (wthrowe) [inline src/DataStructures/DataBox/Prefixes.hpp]: typename

File: src/DataStructures/DataBox/Prefixes.hpp (hunk 5 above)
Comment (wthrowe) [reply src/DataStructures/DataBox/Prefixes.hpp]: I agree, I like 2.

File: src/Domain/Tags.hpp [hunk 7]
```diff
@@ -16,7 +16,27 @@
...
+struct Logical;
+struct Inertial;
+}  // namespace Frame
+
 namespace Tags {
+/// \ingroup DataBoxTagsGroup
+/// \ingroup ComputationalDomainGroup
+/// The logical coordinates in the Element
+template <size_t Dim>
+struct LogicalCoordinates : db::DataBoxTag {
```
Comment (wthrowe) [inline src/Domain/Tags.hpp]: Can this be a compute item based on the extents?

File: src/Domain/Tags.hpp [hunk 8]
```diff
@@ -46,6 +66,23 @@ struct ElementMap : db::DataBoxTag {
...
 };
 
+/// \ingroup DataBoxTagsGroup
+/// \ingroup ComputationalDomainGroup
+/// Computes the inverse Jacobian of the map held by `MapTag` at the coordinates
+/// held by `SourceCoordsTag`. The coordinates must be in the source frame of
+/// the map.
+template <typename MapTag, typename SourceCoordsTag>
+struct InverseJacobian : db::ComputeItemTag {
+  static constexpr db::DataBoxString label = "InverseJacobian";
```
Comment (wthrowe) [inline src/Domain/Tags.hpp]: Can this be a prefix tag so we don't get a name conflict once we have multiple maps?

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp [hunk 9]
```diff
@@ -368,3 +371,162 @@ SPECTRE_TEST_CASE("Unit.Numerical.LinearOperators.PartialDerivs",
...
+  auto box =
+      db::create<db::AddTags<Tags::Extents<Dim>, Tags::Variables<vars_tags>,
+                             Tags::LogicalCoordinates<Dim>, map_tag>,
+                 db::AddComputeItemsTags<inv_jac_tag, deriv_tag>>(
+          extents, u, x_logical, map);
+
+  const auto& du = db::get<deriv_tag>(box);
+
+  for (size_t n = 0; n < du.size(); ++n) {
+    CAPTURE_PRECISE(du.data()[n] - expected_du.data()[n]);  // NOLINT
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp]: Explain NOLINT

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 10]
```diff
@@ -14,32 +16,123 @@
...
+  static constexpr auto function =
+      partial_derivatives<tmpl::list<DerivativeTags...>, variables_tags,
+                          derivative_frame_index::dim,
+                          typename derivative_frame_index::Frame>;
+  using argument_tags =
+      typelist<Tags::Variables<variables_tags>,
+               Tags::Extents<derivative_frame_index::dim>, InverseJacobianTag>;
+};
+
+// Logical partial derivatives
```
Comment (wthrowe) [inline src/DataStructures/DataBox/Prefixes.hpp]: Do we have a known use for derivatives in the logical coordinates?  I expected them to be used to calculate derivatives in other frames, but it looks like we're not doing that.

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 11]
```diff
@@ -14,32 +16,123 @@
...
 };
 
+// Partial derivatives with derivative index in frame related to by
+// InverseJacobianTag
+template <typename... VariablesTags, typename... DerivativeTags,
+          typename InverseJacobianTag>
+struct deriv<tmpl::list<VariablesTags...>, tmpl::list<DerivativeTags...>,
+             InverseJacobianTag,
+             Requires<tt::is_a_v<Tensor, db::item_type<InverseJacobianTag>>>>
+    : db::ComputeItemTag {
```
Comment (wthrowe) [inline src/DataStructures/DataBox/Prefixes.hpp]: This looks like it will give us label conflicts as soon as we have multiple Variables or multiple frames.  I don't know what to do about it here, though, since the prefix tag stuff only uses one parameter.

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 12]
```diff
@@ -14,32 +16,123 @@
...
+template <typename Tag, typename VolumeDim, typename Frame>
+struct deriv<Tag, VolumeDim, Frame,
+             Requires<tt::is_a_v<::Variables, db::item_type<Tag>>>>
     : db::DataBoxPrefix {
   using type = db::item_type<Tag>;
   using tag = Tag;
   static constexpr db::DataBoxString label = "d";
 };
 
+// Partial derivatives with derivative index in frame related to by
```
Comment (wthrowe) [inline src/DataStructures/DataBox/Prefixes.hpp]: related to...the logical coordinates, maybe?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Domain/Tags.hpp (hunk 7 above)
Comment (nilsdeppe) [reply src/Domain/Tags.hpp]: Yep! Good point!

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp (hunk 9 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp]: @kidder why is this NOLINT? I stole it from what you wrote :stuck_out_tongue:

Comment (kidder) [reply tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp]: pointer arithmetic

File: src/Domain/Tags.hpp [hunk 13]
```diff
@@ -32,10 +33,14 @@ namespace Tags {
...
 /// The logical coordinates in the Element
-template <size_t Dim>
-struct LogicalCoordinates : db::DataBoxTag {
+template <typename ExtentsTag>
+struct LogicalCoordinates : db::ComputeItemTag {
   static constexpr db::DataBoxString label = "LogicalCoordinates";
-  using type = tnsr::I<DataVector, Dim, Frame::Logical>;
+  using argument_tags = tmpl::list<ExtentsTag>;
+  static constexpr auto function = logical_coordinates<ExtentsTag::dim>;
+  // (const Index<Dim>& extents) noexcept {
```
Comment (kidder) [inline src/Domain/Tags.hpp]: remove

File: src/Domain/Tags.hpp [hunk 14]
```diff
@@ -32,10 +33,14 @@ namespace Tags {
 /// \ingroup DataBoxTagsGroup
 /// \ingroup ComputationalDomainGroup
 /// The logical coordinates in the Element
-template <size_t Dim>
-struct LogicalCoordinates : db::DataBoxTag {
+template <typename ExtentsTag>
```
Comment (kidder) [inline src/Domain/Tags.hpp]: hard-code Extents

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 15]
```diff
@@ -15,31 +15,6 @@
 #include "Utilities/TypeTraits.hpp"
 
 namespace Tags {
-/// \ingroup DataBoxTagsGroup
```
Comment (kidder) [inline src/DataStructures/DataBox/Prefixes.hpp:18]: couple of typos in commit message move -> moved and at the end of the

File: src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+#include "NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/MakeArray.hpp"
+
+namespace partial_derivatives_detail {
+template <size_t Dim, typename VariableTags, typename DerivativeTags>
+struct LogicalImpl;
+}  // namespace partial_derivatives_detail
+
+/// \ingroup NumericalAlgorithmsGroup
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp]: remove documentation

File: src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp [hunk 17]
```diff
@@ -6,177 +6,134 @@
...
+ *
+ * There are four variants of this tag that change how the derivatives are
+ * computed depending on which is chosen. The two simplest are the non-compute
+ * item versions for Tensor and Variables tags. These take three template
+ * parameters:
+ * 1. The tag to wrap
+ * 2. The volume dim as a type (e.g. `tmpl::size_t<Dim>`)
+ * 3. The frame of the derivative index in `Tensor`s
+ *
+ * The partial derivatives in a Frame that is not the logical frame as a compute
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp]: The third variant

File: src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp [hunk 18]
```diff
@@ -6,177 +6,134 @@
...
+ * item is requested by specifying the list of tags in the
+ * `Tags::Variables<tmpl::list<Tags...>>`, the list of tags to be
+ * differentiated, and the Tag for inverse Jacobian between the logical frame
+ * and the frame in which the derivative is taken. Note that the derivative Tags
+ * must be a contiguous subset from the head of the `Tags...` pack.
+ *
+ * The partial derivatives in the logical frame can also be computed. In that
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp]: The fourth (or final) variant

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]
